### PR TITLE
Add multi-step

### PR DESCRIPTION
**Pull request by yanliang18**

### CODE REVIEW

File: bedrock.hcl [hunk 1]
```diff
@@ -47,16 +47,56 @@ train {
...
             cpu = "0.5"
             memory = "1G"
         }
+        depends_on = ["generate_features"]
     }
 
     parameters {
         RAW_SUBSCRIBERS_DATA = "gs://bedrock-sample/churn_data/subscribers.gz.parquet"
         RAW_CALLS_DATA = "gs://bedrock-sample/churn_data/all_calls.gz.parquet"
+        PREPROCESSED_DATA = "gs://span-artefacts-production/churn_data/preprocessed"
```
Comment (chrissng) [inline bedrock.hcl]: If we want customers to be able to run churn prediction example, this bucket name cannot be hardcoded. That is to say:
- if the environment is `helloworld-production`
- then the pre-created bucket that has write permissions will be `bdrk-helloworld-models`

File: bedrock.hcl [hunk 2]
```diff
@@ -98,22 +137,58 @@ batch_score {
...
```
Comment (chrissng) [inline bedrock.hcl]: Same as above

File: bedrock.hcl (hunk 1 above)
Comment (yanliang18) [reply bedrock.hcl]: is this something the user has to change manually or any way we can get the environment name and automate?

Comment (chrissng) [reply bedrock.hcl]: To keep it simple, i'd suggest specifying the bucket manually through the UI.

Comment (yanliang18) [reply bedrock.hcl]: done

**Comment by chrissng** [comment]
@yanliang18 Got to trouble you to update the README.md to include instructions on how to specify this `TEMP_DATA_BUCKET` parameter prior to running the training pipeline.

@sweatybridge this multi-step churn prediction example would require specifying a the bucket name as the parameter before launching the training pipeline. It defaults to the bucket for playground so current demos should still work. Please see if this PR looks good to you.

File: bedrock.hcl [hunk 3]
```diff
@@ -98,22 +138,59 @@ batch_score {
...
             }}
         ]
+        resources {
+            cpu = "0.5"
+            memory = "1G"
+        }
+        depends_on = ["preprocess"]
+    }
 
+    step train {
```
Comment (chiewwy) [inline bedrock.hcl]:
```suggestion
    step batch_score {
```

File: bedrock.hcl [hunk 4]
```diff
@@ -47,16 +47,57 @@ train {
...
+        install = ["pip3 install --upgrade pip && pip3 install -r requirements.txt"]
+        script = [{sh = ["python3 train.py"]}]
         resources {
             cpu = "0.5"
             memory = "1G"
         }
+        depends_on = ["generate_features"]
     }
 
     parameters {
```
Comment (chiewwy) [inline bedrock.hcl]: To address @chrissng 's comment regarding parameter passing, perhaps we can just add some comments here?
"Parameters will passed to your scripts as environment variables. They can be overwritten later when you create a run."

File: train.py [hunk 5]
```diff
@@ -52,18 +54,14 @@ def compute_log_metrics(gbm, x_val, y_val):
...
     print("\tSplitting train and validation data")
     x_train, x_val, y_train, y_val = train_test_split(
         model_data[FEATURE_COLS],
         model_data[TARGET_COL],
         test_size=0.2,
     )
-
     print("\tTrain model")
     gbm = lgb.LGBMClassifier(
         num_leaves=NUM_LEAVES,
```
Comment (kesamet) [inline train.py]: Should it be
```
with open("/artefact/train/" + OUTPUT_MODEL_NAME, "wb") as model_file:
        pickle.dump(gbm, model_file)
```

or 
```
with open("/artefact/" + OUTPUT_MODEL_NAME, "wb") as model_file:
        pickle.dump(gbm, model_file)
```

**Comment by yanliang18** [comment]
> @yanliang18 Got to trouble you to update the README.md to include instructions on how to specify this `TEMP_DATA_BUCKET` parameter prior to running the training pipeline.
> 
> @sweatybridge this multi-step churn prediction example would require specifying a the bucket name as the parameter before launching the training pipeline. It defaults to the bucket for playground so current demos should still work. Please see if this PR looks good to you.

Updated README

File: train.py (hunk 5 above)
Comment (yanliang18) [reply train.py]: The first. Artefacts are now stored in subfolders (subfolder name is name of step which generated the model). Will update in documentation.

File: bedrock.hcl (hunk 4 above)
Comment (yanliang18) [reply bedrock.hcl]: modified comments about "parameters"

File: train.py (hunk 5 above)
Comment (kesamet) [reply train.py]: Then you also need to update line 75

Comment (yanliang18) [reply train.py]: Sorry I got confused, when saving the latter is correct. We still save into artefact folder but bedrock automatically stores artefacts in subfolders, so when retrieving the model, we need to open from "/artefact/train/..."

Comment (kesamet) [reply train.py]: this is rather confusing, isn't it?

Comment (yanliang18) [reply train.py]: will reply in your slack thread

File: README.md [hunk 6]
```diff
@@ -10,6 +10,14 @@ See [notebook](./doc/churn_prediction.ipynb)
 ### Data processing and model training flowchart
 ![flowchart](./doc/flow.png)
 
+### Run on Bedrock
+Parameters will passed to your scripts as environment variables. They can be overwritten when you create a pipeline run. When you create a run, overwrite these default parameters.
+
+Name | Default value | Description
+--- | --- | ---
+`TEMP_DATA_BUCKET` | "gs://span-artefacts-production/" | Bucket to store intermediate data between steps. If the environment is `test-production`, you may use `bdrk-test-models` which is the pre-created bucket that has write access. You may also use other buckets with write access.
```
Comment (chrissng) [inline README.md]:
```suggestion
`TEMP_DATA_BUCKET` | "gs://span-temp-production/" | Bucket to store intermediate data between steps. If the environment is `test-production`, you may use `bdrk-test-temp` which is the pre-created bucket that has write access. You may also use other buckets with write access.
```

File: bedrock.hcl [hunk 7]
```diff
@@ -47,16 +47,57 @@ train {
...
         RAW_SUBSCRIBERS_DATA = "gs://bedrock-sample/churn_data/subscribers.gz.parquet"
         RAW_CALLS_DATA = "gs://bedrock-sample/churn_data/all_calls.gz.parquet"
+        TEMP_DATA_BUCKET = "gs://span-artefacts-production/"
```
Comment (chrissng) [inline bedrock.hcl]:
```suggestion
        TEMP_DATA_BUCKET = "gs://span-temp-production/"
```

File: bedrock.hcl [hunk 8]
```diff
@@ -98,22 +138,59 @@ batch_score {
...
```
Comment (chrissng) [inline bedrock.hcl]:
```suggestion
        TEMP_DATA_BUCKET = "gs://span-temp-production/"
```

**Pull request merged**